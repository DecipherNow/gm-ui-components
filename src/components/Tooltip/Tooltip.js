import React, { useState } from "react";
import PropTypes from "prop-types";
import { usePopper } from "react-popper";

import TooltipContent from "./components/TooltipContent";
import TooltipWrap from "./components/TooltipWrap";
import { keen } from "style/theme";
export default function Tooltip({
  children,
  content,
  position,
  hideTooltip,
  tooltipStyle,
  theme,
  popperOptions,
  ...props
}) {
  const [referenceElement, setReferenceElement] = useState(null);
  const [popperElement, setPopperElement] = useState(null);
  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    placement: position,
    ...popperOptions
  });
  const [isVisible, setVisibility] = useState(false);

  // Always hide the tooltip if the content is empty
  if (!content) hideTooltip = true;
  return (
    <>
      <TooltipWrap
        ref={isVisible ? setReferenceElement : null}
        onMouseEnter={() => setVisibility(true)}
        onMouseLeave={() => setVisibility(false)}
        {...props}
      >
        {children}
      </TooltipWrap>

      <TooltipContent
        ref={isVisible ? setPopperElement : null}
        style={{
          zIndex: theme.ZINDEX_TOOLTIP,
          ...styles.popper,
          ...tooltipStyle
        }}
        visible={!hideTooltip && isVisible}
        {...attributes.popper}
      >
        {content}
      </TooltipContent>
    </>
  );
}

Tooltip.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node,
    PropTypes.string,
    PropTypes.number
  ]).isRequired,
  content: PropTypes.string,
  hideTooltip: PropTypes.bool,
  position: PropTypes.oneOf([
    "auto",
    "auto-start",
    "auto-end",
    "top",
    "top-start",
    "top-end",
    "bottom",
    "bottom-start",
    "bottom-end",
    "right",
    "right-start",
    "right-end",
    "left",
    "left-start",
    "left-end"
  ]), // https://popper.js.org/docs/v2/constructors/#placement
  tooltipStyle: PropTypes.object,
  popperOptions: PropTypes.object // https://popper.js.org/docs/v2/constructors/#options
};

Tooltip.defaultProps = {
  position: "bottom",
  theme: keen
};

Tooltip.displayName = "Tooltip";
