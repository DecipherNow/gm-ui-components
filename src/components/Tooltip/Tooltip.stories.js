import React from "react";
import { text, select, boolean, object } from "@storybook/addon-knobs";

import Tooltip from "./Tooltip";

export default { title: "Tooltip", component: Tooltip };

export const defaultStory = () => {
  return (
    <Tooltip
      content={text("content", "Content")}
      position={select("position", ["bottom", "right", "top", "left"])}
      hideTooltip={boolean("hideTooltip")}
      tooltipStyle={object("tooltipStyle")}
    >
      {text("children", "Hover over me")}
    </Tooltip>
  );
};

defaultStory.story = {
  name: "Default",

  parameters: {
    info: {
      text: "A tooltip form input with a label."
    }
  }
};

export const typical = () => {
  return (
    <Tooltip
      content={text("content", "Lorem ipsum dolor sit amet")}
      position={select("position", ["bottom", "right", "top", "left"])}
      hideTooltip={boolean("hideTooltip")}
      tooltipStyle={object("tooltipStyle")}
    >
      {text("children", "Hover over me")}
    </Tooltip>
  );
};

typical.story = {
  name: "Typical",

  parameters: {
    info: {
      text: "A tooltip form input with a label."
    }
  }
};

export const longerContent = () => {
  return (
    <Tooltip
      content={text(
        "content",
        "Lorem ipsum dolor sit amet donec consectatur adispacing"
      )}
      position={select("position", ["top", "right", "bottom", "left"])}
      hideTooltip={boolean("hideTooltip")}
      tooltipStyle={object("tooltipStyle")}
    >
      {text("children", "Hover over me")}
    </Tooltip>
  );
};

longerContent.story = {
  name: "Longer content",

  parameters: {
    info: {
      text: "A tooltip form input with a label."
    }
  }
};
