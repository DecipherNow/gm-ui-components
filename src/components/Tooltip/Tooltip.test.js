import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import { render } from "enzyme";

import Tooltip from "./Tooltip";

function renderTooltip(position) {
  return renderer
    .create(
      <Tooltip content="Hello!" position={position}>
        Hover over me
      </Tooltip>
    )
    .toJSON();
}

describe("Tooltip", () => {
  let tree = renderTooltip("top");

  it("correctly positions tooltip content based on position prop", () => {
    expect(tree).toMatchSnapshot();
    tree = renderTooltip("bottom");
    expect(tree).toMatchSnapshot();
    tree = renderTooltip("left");
    expect(tree).toMatchSnapshot();
    tree = renderTooltip("right");
    expect(tree).toMatchSnapshot();
  });

  it("renders a tooltip div with the correct content", () => {
    const wrapper = render(
      <Tooltip content="Hello!" position={"top"}>
        Hover over me
      </Tooltip>
    );
    expect(wrapper.html().includes("Hello!")).toBe(true);
  });
});
