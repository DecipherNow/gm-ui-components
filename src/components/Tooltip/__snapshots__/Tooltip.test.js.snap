// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip correctly positions tooltip content based on position prop 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bwzfXH kkePiY",
    },
    "children": Array [
      Object {
        "data": "Hover over me",
        "next": Object {
          "attribs": Object {
            "class": "sc-bdVaJa bujbxK",
          },
          "children": Array [
            Object {
              "data": "Hello!",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
      Object {
        "attribs": Object {
          "class": "sc-bdVaJa bujbxK",
        },
        "children": Array [
          Object {
            "data": "Hello!",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "data": "Hover over me",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Tooltip correctly positions tooltip content based on position prop 2`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bwzfXH kkePiY",
    },
    "children": Array [
      Object {
        "data": "Hover over me",
        "next": Object {
          "attribs": Object {
            "class": "sc-bdVaJa fseuWE",
          },
          "children": Array [
            Object {
              "data": "Hello!",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
      Object {
        "attribs": Object {
          "class": "sc-bdVaJa fseuWE",
        },
        "children": Array [
          Object {
            "data": "Hello!",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "data": "Hover over me",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Tooltip correctly positions tooltip content based on position prop 3`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bwzfXH kkePiY",
    },
    "children": Array [
      Object {
        "data": "Hover over me",
        "next": Object {
          "attribs": Object {
            "class": "sc-bdVaJa bQyiGV",
          },
          "children": Array [
            Object {
              "data": "Hello!",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
      Object {
        "attribs": Object {
          "class": "sc-bdVaJa bQyiGV",
        },
        "children": Array [
          Object {
            "data": "Hello!",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "data": "Hover over me",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Tooltip correctly positions tooltip content based on position prop 4`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "sc-bwzfXH kkePiY",
    },
    "children": Array [
      Object {
        "data": "Hover over me",
        "next": Object {
          "attribs": Object {
            "class": "sc-bdVaJa jEDpQx",
          },
          "children": Array [
            Object {
              "data": "Hello!",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
      Object {
        "attribs": Object {
          "class": "sc-bdVaJa jEDpQx",
        },
        "children": Array [
          Object {
            "data": "Hello!",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "data": "Hover over me",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "text",
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
